messaging/src/main/java/io/lantern/messaging/Messaging.kt

```diff
@@ -210,37 +215,44 @@ class Messaging(
         contactId: Model.ContactId,
         displayName: String
     ): Model.Contact {
-        val path = contactId.contactPath
         return cryptoWorker.submitForValue {
-            db.mutate { tx ->
-                val existingContact = tx.get<Model.Contact>(path)
-                val contactBuilder = existingContact?.toBuilder() ?: Model.Contact.newBuilder()
-                    .setContactId(contactId)
-                val isNew = existingContact == null
-                if (isNew) {
-                    contactBuilder.createdTs = now
-                    contactBuilder.messagesDisappearAfterSeconds =
-                        defaultMessagesDisappearAfterSeconds
-                }
-                val contact =
-                    contactBuilder.setContactId(contactId).setDisplayName(displayName).build()
-                tx.put(path, contact)
-                // decrypt any "spam" we received from this Contact prior to adding them
-                db.list<ByteArray>(contact.spamQuery)
-                    .forEach { (spamPath, unidentifiedSenderMessage) ->
-                        cryptoWorker.doDecryptAndStore(unidentifiedSenderMessage)
-                        tx.delete(spamPath)
-                    }
-                if (isNew) {
-                    // send our disappear settings as a kind of "hello", that also makes sure we're in sync on retention period
-                    sendDisappearSettings(
-                        tx,
-                        contact.contactId.id,
-                        contact.messagesDisappearAfterSeconds
-                    )
+            doAddOrUpdateContact(contactId, displayName)
```

Comment:
I moved this logic into its own function so that I can use it on an existing CryptoWorker.submit.